usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2367 @@
+    memset(masterkey, 0, MASTER_KEY_SIZE);
+    memset(fname, 0, PATH_MAX);
+    snprintf(fname, PATH_MAX, "%s/MK_USER", data_store);
+    ret = load_masterkey_00(fname, userpin, masterkey);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("Cannot load old masterkey from MK_USER, ret=%08lX.\n",ret);
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    return ret;
+}
+
+/**
+ *
+ */
+CK_RV load_MK_SO_312(const char *data_store, const char *sopin,
+                     TOKEN_DATA *tokdata, CK_BYTE *masterkey)
+{
+    CK_RV ret;
+
+    /* Get masterkey from MK_SO_312. This also verifies SO PIN is correct */
+    memset(masterkey, 0, MASTER_KEY_SIZE);
+    ret = load_masterkey_312(data_store, "MK_SO_312", sopin, tokdata, masterkey);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("Cannot load masterkey from MK_SO_312, ret=%08lX.\n",ret);
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    return ret;
+}
+
+/**
+ *
+ */
+CK_RV load_MK_USER_312(const char *data_store, const char *userpin,
+                       TOKEN_DATA *tokdata, CK_BYTE *masterkey)
+{
+    CK_RV ret;
+
+    /* Get masterkey from MK_USER_312. This also verifies user PIN is correct */
+    memset(masterkey, 0, MASTER_KEY_SIZE);
+    ret = load_masterkey_312(data_store, "MK_USER_312", userpin, tokdata, masterkey);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("Cannot load masterkey from MK_USER_312, ret=%08lX.\n",ret);
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    return ret;
+}
+
+/**
+ * Loads the NVTOK.DAT and returns the TOKEN_DATA struct.
+ */
+CK_RV load_NVTOK_DAT(const char *data_store, const char *nvtok_name,
+                     TOKEN_DATA *td)
+{
+    char fname[PATH_MAX];
+    struct stat stbuf;
+    int fd;
+    size_t tdlen;
+    FILE *fp = NULL;
+    CK_RV ret;
+
+    /* Check parms */
+    if (!data_store || !nvtok_name || !td) {
+        ret = CKR_ARGUMENTS_BAD;
+        goto done;
+    }
+
+    /* Read the NVTOK.DAT */
+    snprintf(fname, PATH_MAX, "%s/%s", data_store, nvtok_name);
+    fp = fopen((char *) fname, "r");
+    if (!fp) {
+        TRACE_ERROR("Cannot open %s, errno=%s\n", fname, strerror(errno));
+        return CKR_FUNCTION_FAILED;
+    }
+
+    fd = fileno(fp);
+    if ((fstat(fd, &stbuf) != 0) || (!S_ISREG(stbuf.st_mode))) {
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Check if this NVTOK.DAT is old or new */
+    if (stbuf.st_size == sizeof(TOKEN_DATA_OLD)) {
+        /* old data store/pin format */
+        tdlen = sizeof(TOKEN_DATA_OLD);
+    } else if (stbuf.st_size == sizeof(TOKEN_DATA)) {
+        /* new data store/pin format */
+        tdlen = sizeof(TOKEN_DATA);
+    } else {
+        TRACE_ERROR("%s has an invalid size of %ld bytes. Neither old nor new token format.\n",
+                    fname, stbuf.st_size);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Read TOKEN_DATA */
+    ret = fread(td, tdlen, 1, fp);
```

Comment:
td is of type *TOKEN_DATA. Is sizeof(TOKEN_DATA) always >= sizeof(TOKEN_DATA_OLD) ? If not, then you overwrite data after the end of *td.

Reply:
yes, TOKEN_DATA is always bigger than TOKEN_DATA_OLD